test/e2e_node/device-plugin.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e_node
+
+import (
+	"fmt"
+	"os/exec"
+	"time"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/pborman/uuid"
+)
+
+const (
+	testPodNamePrefix = "nvidia-gpu-"
+	// Nvidia driver installation can take upwards of 5 minutes.
+	driverInstallTimeout = 10 * time.Minute
+)
+
+var _ = framework.KubeDescribe("[Feature:GPUDevicePlugin]", func() {
+	f := framework.NewDefaultFramework("device-plugin-gpus-errors")
+
+	// Install NVIDIA driver if cluster is running COS
+	BeforeEach(func() {
+		if !framework.IsNodeRunningCOS(getLocalNode(f)) {
+			Skip("NVIDIA GPU tests are supported only on Container Optimized OS image currently")
+		}
+		if !checkIfNvidiaGPUsExistOnNode() {
+			Skip("Nvidia GPUs do not exist on the node. Skipping test.")
+		}
+
+		framework.Logf("Cluster is running on COS. Proceeding with test")
+
+		if framework.NumberOfNVIDIAGPUs(getLocalNode(f)) == 0 {
+			return
+		}
+
+		installDriver(f)
+
+		// Wait for Nvidia GPUs to be available on nodes
+		Eventually(func() bool { return framework.NumberOfNVIDIAGPUs(getLocalNode(f)) > 0 },
+			driverInstallTimeout*2, time.Second).Should(BeTrue())
+	})
+
+	It("checks that when Kubelet restarts exclusive GPU assignation to pods is kept.", func() {
+		By("Creating one GPU pod on a node with at least two GPUs")
+		n := getLocalNode(f)
+
+		if framework.NumberOfNVIDIAGPUs(getLocalNode(f)) < 2 {
+			Skip("Not enough GPUs to execute this test (at least two needed)")
+		}
+
+		p1 := f.PodClient().CreateSync(makeCudaPauseImage())
+
+		By("Restarting Kubelet and creating another GPU pod")
+		restartKubelet(f)
+
```

Comment:
Here can we also add a case to restart p1 and verifies that it gets the same device after restarts?